modules/engine/src/main/scala/com/gsk/kg/engine/Engine.scala

```diff
@@ -706,8 +782,23 @@ object Engine {
       r: Multiset
   )(implicit
       sc: SQLContext
-  ): M[Multiset] =
-    evaluateConstructPlain(bgp, r).pure[M]
+  ): M[Multiset] = {
+
+    // Forward chaining inference uses CONSTRUCT internally. In this case the number of quads in the BGP is very large
+    // and the number of rows in the Multiset is very small. This is an unusual case compared to the use of CONSTRUCT
+    // in normal queries. That is why the evaluation of CONSTRUCT changes in these cases and needs a small adaptation.
+    val forwardChainingQuads: List[Expr] =
+      forwardChainigOWLAxiomnsAndRDFSEntailments
+        .flatMap { case Construct(bgp, _) =>
+          bgp.quads
+        }
+
+    val isInternalForwardChainingUse: Boolean = bgp.quads
+      .map(forwardChainingQuads.contains(_))
+      .reduce(_ && _)
```

Comment:
For performance reasons we should use a `Set` instead of a `List` when calling `contains`. (Also, there's some typos in val/def names in this section.)

```suggestion
    val forwardChainingQuads: Set[Expr] =
      forwardChainingOWLAxiomsAndRDFSEntailments
        .flatMap { case Construct(bgp, _) =>
          bgp.quads
        }.toSet

    val usingForwardChaining: Boolean = bgp.quads
      .map(forwardChainingQuads.contains)
      .reduce(_ && _)
```